### PR TITLE
Add automatic Python formatter `black`

### PR DESCRIPTION
**Pull request by samuelhwilliams**
Adds the standard [black](https://github.com/psf/black) formatter to the project, configured with a line
length of 120. This is an opinionated formatted intended to provide
consistency across many projects (including, these days, many - if not a
majority - of open source projects), and remove any discussion around
personal preference.

---

Git versions >=2.23 allow configuration of a file that lists revs that
should be ignored for the purposes of `git blame`. Let's create that
file and add the black formatting change rev to it.

To use this file, you should configure it with the following command:

`git config --local blame.ignoreRevsFile .git-blame-ignore-revs`

---

🚨⚠️ This will be deployed automatically all the way to production when you click merge ⚠️🚨

For more information, including how to check this deployment on preview or staging first before it goes to production, see our [team wiki section on deployment](https://github.com/alphagov/notifications-manuals/wiki/Merging-and-deploying#deployment)

### CODE REVIEW

**Review by idavidmcdonald — APPROVED** [review]
As someone who cares extremely little about formatting, I have no issue with us having all formatting enforced so we never need to think about it again.

I don't think we've actually spent much time debating over formatting in the past years which is nice so I don't think this is a deal breaker for us but I think there isn't too much downside of having it. Could I get another dev on the team to give a second approval please though before we make this quite opinionated choice. Maybe @klssmith or @CrystalPea?

Question for you Sam. Does introducing black mean that we no longer need flake8 or isort?

Also I think before merging this we need to do a session in the tech show and tell about it, and also about git pre commits because most of the team aren't using them and I think for this change to be painless for devs we need to share the knowledge of the above.

**Comment by samuelhwilliams** [comment]
> Question for you Sam. Does introducing black mean that we no longer need flake8 or isort?

Flake8 covers a number of things that black does not, so we definitely still want that. Black will reformat individual import lines; isort will make sure that they're all grouped correctly and in alphabetical order. So yes, I think we still want that too.

However you've made a good catch - apparently we should also set isort to run with a `black-compatible` profile, so I've added that in now.

**Review by CrystalPea — COMMENTED** [review]
Similar to David, I don't really care that much, and as he said, we never really had formatting wars at Notify. At the same time I don't see any downsides of having this, so if it works better for you @samuelhwilliams then why not.

The only thing that I'd like to clarify, as the language used was unclear to me, is the thing about git blame. Does it mean initial sweeping changes made by Black will be ignored when we look at change history, so that we actually see the more essential changes when we use git blame? If so, will that also work in github?

Otherwise it is a bit of a hassle that for most of the code most recent change will be a formatting change, but not a major problem.

File: app/utils/store.py
Comment (quis) [inline app/utils/store.py:72]: This kind of thing is clearly worse and less readable, but I guess that’s the price you pay for absolute consistency with no human judgment 🙃

Comment (samuelhwilliams) [reply app/utils/store.py:72]: Fair, this is marginally worse, I agree.

**Comment by samuelhwilliams** [comment]
> The only thing that I'd like to clarify, as the language used was unclear to me, is the thing about git blame. Does it mean initial sweeping changes made by Black will be ignored when we look at change history, so that we actually see the more essential changes when we use git blame? If so, will that also work in github?

That's right, yes - the new `.git-blame-ignore-revs` file means that `git blame` should just entirely bypass the mass reformat revision when showing changes to the file. It looks like GitHub also supports this as of this year: https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/. Here's an example of me browsing the latest commit on this branch in blame view:

image

**Comment by klssmith** [comment]
I'd like to check what the change to `git log -p -S` would be before this is merged since, as discussed on Slack, it seems like specific commits can be excluded from `git blame` but not from `git log`. This change would not be worth it for me if we lose the ability to search the repo history using `log`, since that is something I use that a lot.

**Comment by samuelhwilliams** [comment]
@klssmith sure - although if it works how I think it does, and if I was testing the same thing that you use, this shouldn't affect it, as I believe git pickaxe only shows commits that change the number of references of the search string. Which black won't do.

But let me know 👍

**Comment by idavidmcdonald** [comment]
@klssmith @samuelhwilliams did you reach an answer on the behaviour of how this interacts with `git log` etc?

**Comment by klssmith** [comment]
I've tried this out on notifications-admin (since that repo is a lot bigger) and the Black commit wasn't picked up when using git pickaxe as long as you're searching for the contents of a string. That's fine for how I use it, because I wouldn't use it to search for a line - usually only for a variable.

**Pull request merged**
